### PR TITLE
decimal truncation prevention

### PR DESCRIPTION
**Pull request by bcoca**
##### ISSUE TYPE

- Bugfix Pull Request

### CODE REVIEW

File: lib/ansible/config/manager.py [hunk 1]
```diff
@@ -101,7 +101,10 @@ def ensure_type(value, value_type, origin=None, origin_ftype=None):
             value = boolean(value, strict=False)
 
         elif value_type in ('integer', 'int'):
-            value = int(value)
+            if isinstance(value, float) or isinstance(value, string_types) and not value.lstrip('-').isnumeric():
```
Comment (webknjaz) [inline lib/ansible/config/manager.py]: Wouldn't `str.isdigit()` be more appropriate here?
https://docs.python.org/3/library/stdtypes.html#str.isnumeric says that Unicode chars like https://www.compart.com/en/unicode/U+2155 would also match, which is probably not what we want.

File: lib/ansible/playbook/base.py [hunk 2]
```diff
@@ -440,7 +440,15 @@ def get_validated_value(self, name, attribute, value, templar):
...
             value = to_text(value)
         elif attribute.isa == 'int':
-            value = int(value)
+            msg = ''
+            if isinstance(value, string_types) and value.lstrip('-').isnumeric():
+                value = int(value)
+            elif isinstance(value, float):
+                msg = "a 'float', which we cannot cast as it would truncate"
+            else:
+                msg = f"'{value}' of type {type(value)} which is not cleanly castable into an int"
```
Comment (webknjaz) [inline lib/ansible/playbook/base.py]: This produces a confusing `The keyword 'throttle' is required to be of type 'int', but got '0' of type <class 'int'> which is not cleanly castable into an int`: *https://dev.azure.com/ansible/ansible/_build/results?buildId=121618&view=logs&j=bb77140b-c48a-59e0-3d98-c138ebdce3df&t=40ae2d96-445f-5435-416a-957d0826a183&l=559*

Comment (bcoca) [reply lib/ansible/playbook/base.py]: yep, i missed the 'its already an int!' case

File: lib/ansible/config/manager.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/config/manager.py]: yes

File: lib/ansible/playbook/task.py [hunk 3]
```diff
@@ -72,7 +72,7 @@ class Task(Base, Conditional, Taggable, CollectionSearch, Notifiable, Delegatabl
 
     async_val = NonInheritableFieldAttribute(isa='int', default=0, alias='async')
     changed_when = NonInheritableFieldAttribute(isa='list', default=list)
-    delay = NonInheritableFieldAttribute(isa='int', default=5)
+    delay = NonInheritableFieldAttribute(isa='float', default=5)
```
Comment (mattclay) [inline lib/ansible/playbook/task.py]: Changes to existing field types should probably be in a separate PR from changes to field type handling, especially if we're changing to a field type that isn't getting any new behavior.

File: lib/ansible/playbook/base.py [hunk 4]
```diff
@@ -440,7 +440,17 @@ def get_validated_value(self, name, attribute, value, templar):
         if attribute.isa == 'string':
             value = to_text(value)
         elif attribute.isa == 'int':
-            value = int(value)
+            msg = ''
+            if isinstance(value, integer_types):
```
Comment (mattclay) [inline lib/ansible/playbook/base.py]: There's no need to use `integer_types` over `int` since we don't support Python 2.

File: lib/ansible/module_utils/common/validation.py [hunk 5]
```diff
@@ -509,7 +509,7 @@ def check_type_int(value):
     if isinstance(value, integer_types):
         return value
 
-    if isinstance(value, string_types):
+    if isinstance(value, string_types) and value.lstrip('-').isdigit():
```
Comment (mattclay) [inline lib/ansible/module_utils/common/validation.py]: What case(s) does the `isdigit` check handle that aren't already covered by the existing code?

File: lib/ansible/playbook/task.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/playbook/task.py]: agreed, i mostly did it to get ci_complete to find other issues, we already have tests and playbooks in the wild that assume it behaves as a float, so I'll push that sep

File: lib/ansible/module_utils/common/validation.py (hunk 5 above)
Comment (bcoca) [reply lib/ansible/module_utils/common/validation.py]: `0.1`

Comment (mattclay) [reply lib/ansible/module_utils/common/validation.py]: `int('0.1')` raises a `ValueError`, so is already covered by this code, is it not?

File: lib/ansible/playbook/base.py [hunk 6]
```diff
@@ -440,7 +440,17 @@ def get_validated_value(self, name, attribute, value, templar):
...
+                pass  # nothing to do here
+            elif isinstance(value, float):
+                msg = "a 'float', which we cannot cast as it would truncate"
+            elif isinstance(value, string_types) and value.lstrip('-').isdigit():
+                value = int(value)
+            else:
+                msg = f"'{value}' of type {type(value)} which is not cleanly castable into an int"
+            if msg:
+                raise AnsibleParserError(f"The keyword '{name}' is required to be of type 'int', but got {msg}", obj=self.get_ds())
```
Comment (nitzmahone) [inline lib/ansible/playbook/base.py]: Thinking about it a little more, we probably only want to fail on floating-point values that actually have a mantissa. Otherwise, this change will needlessly break anyone feeding one of these keys from, eg, a naive templated division operation (since the result of Python `/` is always `float`, regardless of input). 

There are a number of scenarios where the proposed impl will raise `ValueError`- something like this should cover all the zero-mantissa and very large value cases nicely, while still failing gracefully when a truncation would occur:

```suggestion
            try:
                if not isinstance(value, int):
                    if (decimal_value := decimal.Decimal(value)) != (int_value := int(decimal_value)):
                        raise ValueError(f'Floating-point value {value!r} would be truncated.')
        
                    value = int_value
            except Exception as ex:
                raise AnsibleParserError(f"Keyword {name!r} failed to convert {value!r} to an integer. {ex}", obj=self.get_ds()) from ex

            return value
```

**Review by nitzmahone — COMMENTED** [review]
Should probably only fail on the case where we're actually truncating a mantissa...

File: lib/ansible/module_utils/common/validation.py (hunk 5 above)
Comment (bcoca) [reply lib/ansible/module_utils/common/validation.py]: ah, was testing `int(0.1)` but that is not a string, in any case nitz's code with 'decimal' tracking seems a better fit.

File: lib/ansible/module_utils/common/validation.py [hunk 7]
```diff
@@ -506,17 +506,12 @@ def check_type_int(value):
...
-
-    if isinstance(value, string_types):
-        try:
-            return int(value)
-        except ValueError:
-            pass
-
-    raise TypeError('%s cannot be converted to an int' % type(value))
-
+    if not isinstance(value, int):
```
Comment (mattclay) [inline lib/ansible/module_utils/common/validation.py:509]: Why is this function being changed? The existing behavior strictly requires an int, so there's no risk of truncation. The new logic is more relaxed, allowing values it previously did not accept.

Comment (bcoca) [reply lib/ansible/module_utils/common/validation.py:509]: i was making it consistent with the rest, it will allow things that are 'cleanly' castable to int as do the other two. so `1.1` is still an error but `1.0` is not anymore.

File: lib/ansible/config/manager.py [hunk 8]
```diff
@@ -101,10 +102,14 @@ def ensure_type(value, value_type, origin=None, origin_ftype=None):
             value = boolean(value, strict=False)
 
         elif value_type in ('integer', 'int'):
-            value = int(value)
-
+            if not isinstance(value, int):
+                if (decimal := decimal.Decimal(value)) == (int_part := int(decimal)):
```
Comment (mattclay) [inline lib/ansible/config/manager.py]:
```suggestion
                if (decimal_value := decimal.Decimal(value)) == (int_part := int(decimal_value)):
```

File: lib/ansible/config/manager.py [hunk 9]
```diff
@@ -101,10 +102,14 @@ def ensure_type(value, value_type, origin=None, origin_ftype=None):
             value = boolean(value, strict=False)
 
         elif value_type in ('integer', 'int'):
-            value = int(value)
-
+            if not isinstance(value, int):
+                if (decimal_value := decimal.Decimal(value)) == (int_part := int(decimal_value)):
```
Comment (mattclay) [inline lib/ansible/config/manager.py]: This can fail with an `decimal.InvalidOperation` exception, but the caller (`get_config_value_and_origin`) is only expecting a `ValueError`.

Comment (mattclay) [reply lib/ansible/config/manager.py]: The same issue probably exists in the other locations this pattern is used.

Comment (bcoca) [reply lib/ansible/config/manager.py]: used the underlying decimal.DecimalException, now just need to add tests.

Thought bout just having all use the same function, but I'll leave that for another PR to start consolidating the 'type casting/validation' systems

File: lib/ansible/playbook/base.py [hunk 10]
```diff
@@ -440,7 +441,13 @@ def get_validated_value(self, name, attribute, value, templar):
         if attribute.isa == 'string':
             value = to_text(value)
         elif attribute.isa == 'int':
-            value = int(value)
+            if not isinstance(value, int):
+                try:
+                    if (decimal_value := decimal.Decimal(value)) != (int_value := int(decimal_value)):
+                        raise decimal.DecimalException(f'Floating-point value {value!r} would be truncated.')
+                    value = int_value
+                except decimal.InivalidOperation as e:
```
Comment (webknjaz) [inline lib/ansible/playbook/base.py]:
```python-traceback
AttributeError: module 'decimal' has no attribute 'InivalidOperation'
```
